File: 19 November 2022/C++/Check if all levels of two trees are anagrams or not.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

/* A binary tree node has data, pointer to left child
   and a pointer to right child */
struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};
Node* newNode(int val) {
    Node* temp = new Node;
    temp->data = val;
    temp->left = NULL;
    temp->right = NULL;
    return temp;
}
Node* buildTree(string str) {
    // Corner Case
    if (str.length() == 0 || str[0] == 'N') return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;) ip.push_back(str);

    // Create the root of the tree
    Node* root = newNode(stoi(ip[0]));

    // Push the root to the queue
    queue<Node*> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node* currNode = queue.front();
        queue.pop();

        // Get the current node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current node
            currNode->left = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size()) break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current node
            currNode->right = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}


// } Driver Code Ends
//User function Template for C++

class Solution{
    public:
    bool areAnagrams(Node *r1, Node *r2)
    {
        queue <Node*> q1;
        queue <Node*> q2;
        q1.push(r1);
        q1.push(NULL);
        
        q2.push(r2);
        q2.push(NULL);
        vector <int> v1;
        vector <int> v2;
        while(!q1.empty() ){
            Node* t1=q1.front();
            Node* t2=q2.front();
            if(t1)v1.push_back(t1->data);
            if(t2)v2.push_back(t2->data);
            q1.pop();
            q2.pop();
            if(t1==NULL and t2==NULL){
                if((!q1.empty()) and (!q2.empty())){
                    q1.push(NULL);
                    q2.push(NULL);
                }
                sort(v1.begin(),v1.end());
                sort(v2.begin(),v2.end());
                if(v1!=v2) return false;
                v1.clear();
                v2.clear();
            }
            else if(t1==NULL || t2==NULL){
                return false;
            }
            else{
                
                    if(t1->left){
                        q1.push(t1->left);
                    }
                    if(t1->right){
                        q1.push(t1->right);
                    }
                    if(t2->right){
                        q2.push(t2->right);
                    }
                    if(t2->left){
                        q2.push(t2->left);
                    }
                    
               
            }
        }
        return true;
    }
};

//{ Driver Code Starts.

/* Driver program to test size function*/
int main() {
    int t;
    scanf("%d\n", &t);
    while (t--) {
        string treeString1, treeString2;
        getline(cin, treeString1);
        Node *root1 = buildTree(treeString1);
        getline(cin, treeString2);
        Node *root2 = buildTree(treeString2);
        Solution ob;
        cout << ob.areAnagrams(root1, root2) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: 19 November 2022/Java/Check if all levels of two trees are anagrams or not.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;


class Node
{
    int data;
    Node left;
    Node right;
    
    Node(int data)
    {
        this.data = data;
        left = null;
        right = null;
    }
    
    public static Node buildTree(String str)
    {
        // Corner Case
        if(str.length()==0 || str.charAt(0)=='N')
            return null;
        
        // Creating array of Strings from input 
        // String after spliting by space
        String ip[] = str.split(" ");
        
        // Create the root of the tree
        Node root = new Node(Integer.parseInt(ip[0]));
        
        // Push the root to the queue
        Queue<Node> queue = new LinkedList<>(); 
        queue.add(root);
        
        // Starting from the second element
        int i = 1;
        while(queue.size()>0 && i < ip.length)
        {
            
            // Get and remove the front of the queue
            Node currNode = queue.peek();
            queue.remove();
                
            // Get the current node's value from the string
            String currVal = ip[i];
                
            // If the left child is not null
            if(!currVal.equals("N"))
            {
                    
                // Create the left child for the current node
                currNode.left = new Node(Integer.parseInt(currVal));
                // Push it to the queue
                queue.add(currNode.left);
            }
                
            // For the right child
            i++;
            if(i >= ip.length)
                break;
                
            currVal = ip[i];
                
            // If the right child is not null
            if(!currVal.equals("N"))
            {
                    
                // Create the right child for the current node
                currNode.right = new Node(Integer.parseInt(currVal));
                    
                // Push it to the queue
                queue.add(currNode.right);
            }
            i++;
        }
        
        return root;
    }
    
    public static Node inputTree(BufferedReader br) throws IOException
    {
        return buildTree(br.readLine().trim());
    }
    
    public static void inorder(Node root)
    {
        if (root == null)
           return;
        inorder(root.left);
        System.out.print(root.data + " ");
        inorder(root.right);
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while(t-- > 0){
            
            Node node1 = Node.inputTree(br);
            
            
            Node node2 = Node.inputTree(br);
            
            Solution obj = new Solution();
            boolean res = obj.areAnagrams(node1, node2);
            
            int _result_val = (res) ? 1 : 0;
            System.out.println(_result_val);
        }
    }
}

// } Driver Code Ends


/*

Definition for Binary Tree Node
class Node
{
    int data;
    Node left;
    Node right;
    
    Node(int data)
    {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    public static boolean areAnagrams(Node node1, Node node2) {
        Queue<Node> q1=new LinkedList<>();
        Queue<Node> q2=new LinkedList<>();
        q1.offer(node1);
        q2.offer(node2);
        while(q1.isEmpty()==false && q2.isEmpty()==false)
        {
            int s1=q1.size();
            int s2=q2.size();
            if(s1!=s2)
                return false;
            HashMap<Integer, Integer> map=new HashMap<>();
            for(int i=0;i<s1;i++)
            {
                Node x1=q1.poll();
                Node x2=q2.poll();
                if(x1.left!=null)
                    q1.offer(x1.left);
                if(x2.left!=null)
                    q2.offer(x2.left);
                if(x1.right!=null)
                    q1.offer(x1.right);
                if(x2.right!=null)
                    q2.offer(x2.right);
                map.put(x1.data, map.getOrDefault(x1.data, 0)+1);
                map.put(x2.data, map.getOrDefault(x2.data, 0)-1);
            }
            for(int x: map.values())
            {
                if(x!=0)
                    return false;
            }
        }
        if(q1.isEmpty()==false || q2.isEmpty()==false)
        {
            return false;
        }
        return true;
    }
}
        
